protons/driver.py

```diff
@@ -813,45 +916,43 @@ def _attempt_state_change(self, context, reject_on_nan=False):
                 # Use instantaneous switching.
                 for titration_group_index in titration_group_indices:
                     self._set_titration_state(titration_group_index, final_titration_states[titration_group_index], context)
-            else:
-                # Run NCMC integration.
-                for step in range(self.nsteps_per_trial):
-                    # Take a Verlet integrator step.
-                    self.ncmc_propagation_integrator.step(1)
-                    # Update the sams_sampler state.
-                    titration_lambda = float(step + 1) / float(self.nsteps_per_trial)
-                    # TODO: Using a VerletIntegrator together with half-kicks on either side would save one force evaluation per iteration,
-                    # since parameter update would occur in the middle of a velocity Verlet step.
-                    # TODO: This could be optimized by only calling
-                    # context.updateParametersInContext once rather than after every sams_sampler
-                    # state update.
-                    for titration_group_index in titration_group_indices:
-                        self._update_forces(titration_group_index, final_titration_states[titration_group_index], initial_titration_state_index=initial_titration_states[
-                                            titration_group_index], fractional_titration_state=titration_lambda, context=context)
-                        # TODO: Optimize where integrator.step() is called
-                    self.ncmc_propagation_integrator.step(1)
-
-                # Update sams_sampler states so that log state penalties are accurately reflected.
-                for titration_group_index in titration_group_indices:
-                    self.titrationStates[titration_group_index] = titration_state_index
 
-            # Compute final probability of this protonation state.
-            log_P_final, pot2, kin2 = self._compute_log_probability(context)
+                log_P_final, pot2, kin2 = self._compute_log_probability(context)
+                work = - (log_P_final - log_P_initial)
+            else:
+                # Only perform NCMC when the proposed state is different from the current state
+                if initial_titration_states[titration_group_index] != final_titration_states[titration_group_index]:
+                    # Run NCMC integration.
+                    work = self._ncmc_ghmc(context, titration_group_indices, initial_titration_states, final_titration_states)
+                else:
+                    work = 0.0
+                # Save the kinetic and potential energy for records
+                log_P, pot2, kin2 = self._compute_log_probability(context)
+
+            #TODO: remove when certain velocity Verlet won't be used as the NCMC propagator
```

Comment:
We should be able to just remove this. If we need it, we can retrieve it from the git history.